app/src/org/commcare/dalvik/activities/PinAuthenticationActivity.java

```diff
@@ -0,0 +1,173 @@
+package org.commcare.dalvik.activities;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import org.commcare.android.database.app.models.UserKeyRecord;
+import org.commcare.android.framework.ManagedUi;
+import org.commcare.android.framework.SessionAwareCommCareActivity;
+import org.commcare.android.framework.UiElement;
+import org.commcare.dalvik.R;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * When a user selects Set/Reset PIN from the home activity options menu, they must first
```

Comment:
Worth noting that depending upon what you propose as an alternative to passing the un-hashed password through from login (in place of doing it via the intent), it's possible that we could still have it around from the initial login. Although it seems like that wouldn't be a great situation, and we should probably wipe the password from wherever we cache it as soon as possible no matter what, instead of allowing it to persist indefinitely just so the user doesn't have to re-auth here.